### ISSUE TITLE
[POLL][RFC] Can we retire Single-Process Multi-Device Mode from DistributedDataParallel?

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#47012]
## Update 03/22/2021

@SciPioneer started working on retiring SPMD and it's planned to be removed in v1.9. Please let us know if you have concerns. 

---------------------------

## Background

As of today, applications can use either **Single-Process Multi-Device** (SPMD) mode or **Single-Process Single-Device** (SPSD) mode in DDP by setting `device_ids` argument in DDP's constructor. E.g.:

```python
# implicit SPMD mode: when device_ids is missing, DDP will use all visible devices
ddp_model = DistributedDataParallel(model)

# explicit SPMD mode: by specifying devices
ddp_model = DistributedDataParallel(model, device_ids=[d1, d2, d3, d4])

#SPSD mode
ddp_model = DistributedDataParallel(model, device_ids=[d1])
```

Since v1.5, we added a warning to highlight that SPMD mode in DDP is not recommend, and we suggest using Single-Process Single-Device (SPSD) mode, where each process only operates on one model replica. 

https://github.com/pytorch/pytorch/blob/14d87ec5a345891217dc77a88a5ec29272909383/torch/nn/parallel/distributed.py#L461-L471

Since then, new features like DDP `join`, DDP communication hook, etc. are not targeting SPMD mode any more. Now, we are at v1.7, and we have only seen a small number of concerns (#36268) on that warning so far, which will be addressed below. So I wonder if we can kick off the process to retire SPMD from DDP.

In this RFC, we would like to collect feedback on whether we can safely retire the SPMD from DDP. The deprecation will help simplify DDP implementation and also help avoid maintaining similar logic in both `DataParallel` and `DistributedDataParallel`. 

**Note that:**

1. This deprecation will **NOT** happen anytime soon, but we would like to understand the gaps earlier to act accordingly. 
2. This RFC will **NOT** touch `DataParallel`. cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @jiayisuse @agolynski @VitalyFedyunin 

## Pitch

1. Let DDP error out if `device_ids` contains multiple devices.
2. If `device_ids` is not specified, DDP will use the provided model (`module` argument in DDP constructor) as-is, regardless if the model is on one GPU or multiple GPUs or on CPU.

**Please do let us know if you have concerns.**

cc @meixitu @leimao @adrianosantospb 

## Motivation

1. SPMD mode in DDP is slow. 
2. SPMD mode introduces additional complexity to DDP implementation, while most applications don't use it.
3. SPMD mode in DDP on a single machine can be replaced by `DataParallel`, which is also easier to setup compared to DDP.

### NLP Applications

See the context #36268. IIUC, the NLP use case mentioned there uses SPMD DDP because it needs to have a large number of negative samples. By gathering outputs from multiple GPUs, the aggregated output will contain more negative samples. In contrast, when using SPSD DDP, the negative samples from one GPU will be too few, which might hurt the quality of local loss? (Please correct me if I am wrong. cc @williamFalcon @stephenroller @zhangguanheng66 )

**Solution:**

I see multiple solutions for this problem:

1. We can add an autograd function for `dist.gather` (WIP PR: #40762), and then SPSD DDP can call `dist.gather` accordingly to gather outputs. 
2. We can use RPC together with DDP, and collect outputs from multiple processes through RPC. There might be gaps here, but we can work on closing those gaps and provide a tutorial to help the migration. The application code can be sth like:

```python
with dist_autograd.context() as ctx_id:
	fut_outs = []
	for ddp in ddp_modules:
		fut_outs.append(ddp.rpc_async().forward(inps))

	gathered_outs = torch.concat(torch.futures.wait_all(fut_outs))
	dist.backward(ctx_id, [loss_fn(gathered_outs)])
```

**Comment by stephenroller** [comment]
The ranking model distributed setting (where the many negatives benefits) has been de-prioritized in ParlAI due to advancements in generative models. It's acceptable for me to lose that functionality now.

**Comment by stephenroller** [comment]
Also, single process pipeline model parallelism can let us achieve the same result, as transformers are easy to mix with GPipe/Megatron parallelism these days, which works well with DDP as is.

**Comment by williamFalcon** [comment]
cc @SeanNaren

**Comment by zhangguanheng66** [comment]
My understanding is that SPMD mode in DDP is similar to `DataParallel`. Before retiring the SPMD mode, should we provide a mitigation example and show those users how to transfer to `DataParallel` (if they really want to).

**Comment by meixitu17** [comment]
I always use SPSD after I know this skill

**Comment by VitalyFedyunin** [comment]
> My understanding is that SPMD mode in DDP is similar to `DataParallel`. Before retiring the SPMD mode, should we provide a mitigation example and show those users how to transfer to `DataParallel` (if they really want to).

No please. We are looking to release very-long-term DataParallel deprecation plan (with reasons). Promoting it as replacement for SPMD is bad idea.

**Comment by mrshenli** [comment]
Hey @zhangguanheng66, is there any torchtext application that is using SPMD DDP and might be hard to migrate to SPSD DDP?

**Comment by zhangguanheng66** [comment]
> Hey @zhangguanheng66, is there any torchtext application that is using SPMD DDP and might be hard to migrate to SPSD DDP?

In the BERT model, we use SPSD DDP so SPMD is not a concern for torchtext.

**Comment by zhangguanheng66** [comment]
> > My understanding is that SPMD mode in DDP is similar to `DataParallel`. Before retiring the SPMD mode, should we provide a mitigation example and show those users how to transfer to `DataParallel` (if they really want to).
> 
> No please. We are looking to release very-long-term DataParallel deprecation plan (with reasons). Promoting it as replacement for SPMD is bad idea.

That's why I said "if they really want to". I agree we shouldn't promote DataParallel as a replacement but the current SPMD users might need a backup plan for this.

**Comment by mcarilli** [comment]
Does anyone remember a few months ago when I discovered SPMD DDP was unusably broken?  The fact that no one reported it, and my tests were the first to discover it, indicates there were no "current SPMD users."  There could not have been.  It simply didn't work.

**Comment by mrshenli** [comment]
Hey @mcarilli, thanks for the comment. According to the discussion in #36268, it's working for <= v1.4. And your fix make it work for >= v1.6. It does not work for v1.5 is a known issue. So if users jump from v1.4 to v1.6, they probably wouldn't notice that.

**Comment by rohan-varma** [comment]
> > My understanding is that SPMD mode in DDP is similar to `DataParallel`. Before retiring the SPMD mode, should we provide a mitigation example and show those users how to transfer to `DataParallel` (if they really want to).
> 
> No please. We are looking to release very-long-term DataParallel deprecation plan (with reasons). Promoting it as replacement for SPMD is bad idea.

Is there a plan to have an RFC soon for the deprecation of `DataParallel`?
